File: react-native/android/src/main/java/com/digia/engage/rn/DigiaModule.kt
/**
 * DigiaModule
 *
 * React Native NativeModule that bridges the Digia Engage Android SDK.
 *
 * Exposed methods (callable from JS via NativeModules.DigiaEngageModule):
 *
 * initialize(apiKey, environment, logLevel, baseUrl): Promise<void> register(): void
 * setCurrentScreen(name): void triggerCampaign(cepCampaignId, campaignKey, variables, cepMetadata): void
 * invalidateCampaign(campaignId): void
 *
 * Architecture ──────────── The RN bridge mirrors the native Digia.initialize / Digia.register /
 * Digia.setCurrentScreen flow exactly. An internal [RNEventBridgePlugin] is the single native
 * DigiaCEPPlugin registered via Digia.register().
 *
 * When the SDK calls plugin.setup(delegate), the bridge stores that delegate reference. JS plugins
 * that need to push campaigns into the Compose overlay call triggerCampaign / invalidateCampaign
 * which forward to delegate.onCampaignTriggered / delegate.onCampaignInvalidated.
 *
 * Overlay lifecycle events (impressed / clicked / dismissed) are forwarded from the native
 * plugin.notifyEvent() to JS via DeviceEventEmitter so that pure-JS CEP plugins (e.g.
 * DigiaMoEngagePlugin) can report analytics.
 */
package com.digia.engage.rn

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModelStoreOwner
import androidx.lifecycle.setViewTreeLifecycleOwner
import androidx.lifecycle.setViewTreeViewModelStoreOwner
import androidx.savedstate.SavedStateRegistryOwner
import androidx.savedstate.setViewTreeSavedStateRegistryOwner
import com.digia.engage.CEPTriggerPayload
import com.digia.engage.DiagnosticReport
import com.digia.engage.Digia
import com.digia.engage.DigiaCEPDelegate
import com.digia.engage.DigiaCEPPlugin
import com.digia.engage.DigiaConfig
import com.digia.engage.DigiaEnvironment
import com.digia.engage.DigiaExperienceEvent
import com.digia.engage.DigiaHostView
import com.digia.engage.DigiaLogLevel
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.UiThreadUtil
import com.facebook.react.modules.core.DeviceEventManagerModule

internal class DigiaModule(
        private val reactContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(reactContext) {

    private val rnPlugin = RNEventBridgePlugin(reactContext)

    override fun getName(): String = MODULE_NAME

    // ─── initialize ───────────────────────────────────────────────────────────

    @ReactMethod
    fun initialize(
            apiKey: String,
            environment: String,
            logLevel: String,
            baseUrl: String?,
            fontFamily: String?,
            sdkVersion: String?,
            promise: Promise
    ) {
        try {
            val cleanBaseUrl =
                    baseUrl?.trim()?.trimEnd('/')?.removeSuffix("/api/v1")?.takeIf { it.isNotBlank() }
            val config =
                    DigiaConfig(
                            apiKey = apiKey,
                            baseUrl = cleanBaseUrl,
                            environment =
                                    when (environment.lowercase()) {
                                        "sandbox" -> DigiaEnvironment.SANDBOX
                                        else -> DigiaEnvironment.PRODUCTION
                                    },
                            logLevel =
                                    when (logLevel.lowercase()) {
                                        "verbose" -> DigiaLogLevel.VERBOSE
                                        "none" -> DigiaLogLevel.NONE
                                        else -> DigiaLogLevel.ERROR
                                    },
                            fontFamily = fontFamily?.takeIf { it.isNotBlank() },
                            // This native engine is being driven by the RN wrapper.
                            wrapperBinding = "react_native",
                            wrapperVersion = sdkVersion?.takeIf { it.isNotBlank() },
                    )
            Digia.initialize(reactContext.applicationContext, config)

            UiThreadUtil.runOnUiThread {
                // Mount the Compose overlay ABOVE the ReactRootView via addContentView().
                // This keeps it outside Fabric's shadow tree entirely so Fabric hit-testing
                // never sees it. Touch pass-through is handled by DigiaHostView.dispatchTouchEvent
                // returning false at the native Android level — which works correctly at this
                // level of the hierarchy, unlike inside Fabric where pointerEvents="none" on
                // non-ReactViewGroup views is not respected during shadow-tree hit-testing.
                mountDigiaHost()
                promise.resolve(null)
            }
        } catch (e: Exception) {
            promise.reject("DIGIA_INIT_ERROR", e.message ?: "Initialisation failed", e)
        }
    }

    // ─── register ─────────────────────────────────────────────────────────────

    /**
     * Registers [RNEventBridgePlugin] with the native Digia SDK.
     *
     * This is bridge infrastructure — not a user-facing CEP plugin. Called once by the JS
     * `Digia.register()` wrapper on the first plugin registration so that
     * [RNEventBridgePlugin.delegate] is populated before any triggerCampaign / invalidateCampaign
     * calls arrive from JS.
     */
    @ReactMethod
    fun registerBridge() {
        Digia.register(rnPlugin)
    }

    // ─── setCurrentScreen ─────────────────────────────────────────────────────

    @ReactMethod
    fun setCurrentScreen(name: String) {
        Digia.setCurrentScreen(name)
    }

    // ─── Analytics identity ───────────────────────────────────────────────────

    @ReactMethod
    fun setUserId(userId: String) {
        Digia.setUserId(userId)
    }

    @ReactMethod
    fun clearUserId() {
        Digia.clearUserId()
    }

    // ─── Analytics event forwarding (guide / JS-rendered campaigns) ───────────

    /**
     * Records an analytics event from a JS-rendered campaign (guide). [eventName] is
     * the Engage matrix event name; [props] carries the wire-keyed fields. The native
     * SDK maps it to the typed analytics event and records it. Native campaigns
     * (nudge, inline, survey) are tracked internally by the SDK.
     */
    @ReactMethod
    fun captureAnalyticsEvent(campaignKey: String, eventName: String, props: ReadableMap) {
        android.util.Log.d("DigiaAnalytics", "[captureAnalyticsEvent] name='$eventName' campaignKey=$campaignKey")
        Digia.captureAnalyticsEvent(campaignKey, eventName, props.toHashMap().toMap())
    }

    // ─── triggerCampaign ──────────────────────────────────────────────────────

    /**
     * Forwards a campaign payload to the native DigiaCEPDelegate.
     *
     * This is called by the JS DigiaDelegate.onCampaignTriggered() implementation when a JS CEP
     * plugin (e.g. DigiaMoEngagePlugin) delivers a campaign. The delegate routes it into the
     * Compose overlay for rendering.
     */
    @ReactMethod
    fun triggerCampaign(
        cepCampaignId: String,
        campaignKey: String,
        variables: ReadableMap,
        cepMetadata: ReadableMap,
    ) {
        val delegate = rnPlugin.delegate ?: return
        val vars = variables.toHashMap()
            .entries
            .associate { it.key to it.value?.toString().orEmpty() }
            .takeIf { it.isNotEmpty() }
        delegate.onCampaignTriggered(
                CEPTriggerPayload(
                        cepCampaignId = cepCampaignId,
                        campaignKey = campaignKey,
                        cepMetadata = cepMetadata.toHashMap().toMap(),
                        variables = vars,
                )
        )
    }

    // ─── invalidateCampaign ───────────────────────────────────────────────────

    /** Forwards a campaign invalidation to the native DigiaCEPDelegate. */
    @ReactMethod
    fun invalidateCampaign(campaignId: String) {
        rnPlugin.delegate?.onCampaignInvalidated(campaignId)
    }

    // ─── Anchor registration ──────────────────────────────────────────────────

    @ReactMethod
    fun registerAnchor(key: String, x: Int, y: Int, width: Int, height: Int) {
        Digia.registerAnchor(key, x, y, width, height)
    }

    @ReactMethod
    fun unregisterAnchor(key: String) {
        Digia.unregisterAnchor(key)
    }

    @ReactMethod
    fun getRegisteredComponents(promise: Promise) {
        promise.resolve(Arguments.createArray())
    }

    // ─── Internal: mount the Compose overlay host ─────────────────────────────

    private fun mountDigiaHost() {
        val activity =
                reactContext.currentActivity
                        ?: run {
                            android.util.Log.w(
                                    "DigiaHost",
                                    "[mountDigiaHost] no current activity — skipping"
                            )
                            return
                        }

        val contentRoot = activity.window.decorView.findViewWithTag<DigiaHostView>(DIGIA_HOST_TAG)
        if (contentRoot != null) {
            android.util.Log.d(
                    "DigiaHost",
                    "[mountDigiaHost] already mounted (tag found) — skipping"
            )
            return
        }

        android.util.Log.d("DigiaHost", "[mountDigiaHost] mounting native overlay on DecorView")

        // Mount on the DecorView (not content area) so the overlay covers the full screen
        // including status bar and navigation bar. This also ensures the Compose Popup's
        // canvas y=0 aligns with the absolute screen y=0, matching getLocationOnScreen()
        // coordinates used by DigiaAnchorView.
        val decorView =
                activity.window.decorView as? android.view.ViewGroup
                        ?: run {
                            android.util.Log.w(
                                    "DigiaHost",
                                    "[mountDigiaHost] decorView not a ViewGroup — skipping"
                            )
                            return
                        }

        val composeView =
                DigiaHostView(activity).apply {
                    tag = DIGIA_HOST_TAG
                    if (activity is LifecycleOwner) setViewTreeLifecycleOwner(activity)
                    if (activity is ViewModelStoreOwner) setViewTreeViewModelStoreOwner(activity)
                    if (activity is SavedStateRegistryOwner)
                            setViewTreeSavedStateRegistryOwner(activity)
                }

        decorView.addView(
                composeView,
                android.view.ViewGroup.LayoutParams(
                        android.view.ViewGroup.LayoutParams.MATCH_PARENT,
                        android.view.ViewGroup.LayoutParams.MATCH_PARENT,
                ),
        )
        android.util.Log.d(
                "DigiaHost",
                "[mountDigiaHost] done — DecorView child count: ${decorView.childCount}"
        )
    }

    companion object {
        const val MODULE_NAME = "DigiaEngageModule"
        private const val DIGIA_HOST_TAG = "digia_host_compose_view"
    }
}

/**
 * RNEventBridgePlugin
 *
 * The single native DigiaCEPPlugin used in React Native.
 *
 * When Digia.register(rnPlugin) is called, the SDK calls setup(delegate) which gives us the
 * DigiaCEPDelegate reference. This delegate is used by triggerCampaign / invalidateCampaign bridge
 * methods to push campaigns into the native rendering engine.
 *
 * Overlay lifecycle events (impressed / clicked / dismissed) received via notifyEvent() are emitted
 * to JS as DeviceEventEmitter events so that JS CEP plugins can report analytics back to their
 * platform.
 */
internal class RNEventBridgePlugin(
        private val reactContext: ReactApplicationContext,
) : DigiaCEPPlugin {

    override val identifier: String = "rn-event-bridge"

    /** Delegate received from the SDK via setup(). Used by DigiaModule bridge methods. */
    var delegate: DigiaCEPDelegate? = null
        private set

    private fun emit(event: String, params: com.facebook.react.bridge.WritableMap) {
        if (reactContext.hasActiveReactInstance()) {
            reactContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit(event, params)
        }
    }

    override fun setup(delegate: DigiaCEPDelegate) {
        this.delegate = delegate
    }

    override fun forwardScreen(name: String) {
        /* forwarded by Digia.setCurrentScreen() on the native side */
    }

    override fun notifyEvent(event: DigiaExperienceEvent, payload: CEPTriggerPayload) {
        val params =
                Arguments.createMap().apply {
                    putString("campaignId", payload.cepCampaignId)
                    when (event) {
                        is DigiaExperienceEvent.Impressed -> putString("type", "impressed")
                        is DigiaExperienceEvent.Clicked -> {
                            putString("type", "clicked")
                            event.elementId?.let { putString("elementId", it) }
                        }
                        is DigiaExperienceEvent.Dismissed -> putString("type", "dismissed")
                    }
                }
        emit("digiaEngageEvent", params)
    }

    override fun notifyAction(actionType: String, url: String, payload: CEPTriggerPayload): Boolean {
        val params =
                Arguments.createMap().apply {
                    putString("campaignId", payload.cepCampaignId)
                    putString("type", "action")
                    putString("actionType", actionType)
                    putString("url", url)
                }
        emit("digiaEngageEvent", params)
        return true
    }

    override fun teardown() {
        delegate = null
    }

    override fun healthCheck(): DiagnosticReport =
            DiagnosticReport(isHealthy = true, metadata = mapOf("identifier" to identifier))
}

File: react-native/ios/DigiaModule.swift
import DigiaEngage
/**
 * DigiaModule
 *
 * React Native NativeModule that bridges the Digia Engage iOS SDK.
 *
 * Exposed methods (callable from JS via NativeModules.DigiaEngageModule):
 *   initialize(projectId, environment, logLevel, baseUrl): Promise<void>
 *   registerBridge(): void
 *   setCurrentScreen(name): void
 *   triggerCampaign(id, content, cepContext): void
 *   invalidateCampaign(campaignId): void
 *   createInitialPage(): void  // AppConfig initial route
 *
 * Architecture
 * ────────────
 * The RN bridge mirrors the native Digia.initialize / Digia.register /
 * Digia.setCurrentScreen flow exactly. An internal RNEventBridgePlugin is
 * the single native DigiaCEPPlugin registered via Digia.register().
 *
 * When the SDK calls plugin.setup(delegate:), the bridge stores that
 * delegate reference. JS plugins that deliver campaigns call triggerCampaign /
 * invalidateCampaign which forward to delegate.onCampaignTriggered /
 * delegate.onCampaignInvalidated.
 *
 * Overlay lifecycle events (impressed / clicked / dismissed) are forwarded from
 * the native plugin.notifyEvent() to JS via RCTEventEmitter so that pure-JS
 * CEP plugins (e.g. DigiaMoEngagePlugin) can report analytics.
 */
import Foundation
import React
import SwiftUI
import UIKit

@objc(DigiaEngageModule)
final class DigiaModule: RCTEventEmitter {

    private lazy var rnPlugin: RNEventBridgePlugin = MainActor.assumeIsolated {
        RNEventBridgePlugin(eventEmitter: self)
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - RCTEventEmitter

    override static func requiresMainQueueSetup() -> Bool { true }

    override init() {
        super.init()
        // Pre-seed _listenerCount = 1 so sendEventWithName: never silently drops
        // events when JS uses DeviceEventEmitter (which doesn't call native
        // addListener: on iOS and therefore never increments the count).
        addListener("digiaEngageEvent")
    }

    override func supportedEvents() -> [String]! {
        return ["digiaEngageEvent"]
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - initialize

    @objc
    func initialize(
        _ projectId: String,
        environment: String,
        logLevel: String,
        baseUrl: String?,
        fontFamily: String?,
        resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        let envValue: DigiaEnvironment =
            environment.lowercased() == "sandbox" ? .sandbox : .production
        let logLevelValue: DigiaLogLevel
        switch logLevel.lowercased() {
        case "verbose": logLevelValue = .verbose
        case "none": logLevelValue = .none
        default: logLevelValue = .error
        }

        let cleanBaseUrl = baseUrl.flatMap { url -> String? in
            var s = url.trimmingCharacters(in: .whitespacesAndNewlines)
                .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
            if s.hasSuffix("/api/v1") { s = String(s.dropLast(7)) }
            return s.isEmpty ? nil : s
        }

        let config = DigiaConfig(
            apiKey: projectId,
            logLevel: logLevelValue,
            environment: envValue,
            developerConfig: cleanBaseUrl.map { DigiaDeveloperConfig(baseURL: $0) },
            fontFamily: fontFamily.flatMap { $0.isEmpty ? nil : $0 }
        )

        Task { @MainActor in
            do {
                try await Digia.initialize(config)
                self.mountDigiaHost()
                resolve(nil)
            } catch {
                reject("DIGIA_INIT_ERROR", error.localizedDescription, error)
            }
        }
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - registerBridge

    /// Registers RNEventBridgePlugin with the native Digia SDK.
    /// Called automatically by the JS Digia.register() wrapper on first plugin
    /// registration so that the delegate is populated before any
    /// triggerCampaign / invalidateCampaign calls arrive from JS.
    @objc
    func registerBridge() {
        Task { @MainActor in
            // Dismiss any stale overlay left over from the previous JS session.
            Digia.dismissActiveNudge()

            // After a JS reload, re-bring the overlay host to the front.
            // Expo/RN may have added views during the reload cycle that sit on
            // top of the host, causing the SwiftUI layer to be unreachable by
            // touch even when hasActiveOverlay = true.
            if let rootVC = UIApplication.shared.connectedScenes
                .compactMap({ ($0 as? UIWindowScene)?.keyWindow?.rootViewController })
                .first,
               let hostView = rootVC.view.viewWithTag(Self.overlayMountTag) {
                rootVC.view.bringSubviewToFront(hostView)
            }

            Digia.register(self.rnPlugin)
            print("[DigiaRN] registerBridge: rnPlugin registered, delegate=\(self.rnPlugin.delegate != nil ? "set" : "nil")")
        }
    }

    private static let overlayMountTag = 0xD19140

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - setCurrentScreen

    @objc
    func setCurrentScreen(_ name: String) {
        // Digia.setCurrentScreen is not yet part of the public iOS API; this is
        // a no-op placeholder that can be activated once it is added.
        // Digia.setCurrentScreen(name)
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - Analytics identity

    @objc
    func setUserId(_ userId: String) {
        Task { @MainActor in Digia.setUserId(userId) }
    }

    @objc
    func clearUserId() {
        Task { @MainActor in Digia.clearUserId() }
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - trackEvent (guide / JS-rendered campaigns)

    @objc
    func trackEvent(
        _ eventType: String,
        campaignId: String,
        campaignKey: String,
        campaignType: String,
        elementId: String?
    ) {
        print("[DigiaAnalytics] [DigiaRN] trackEvent type=\(eventType) campaignId=\(campaignId) campaignKey=\(campaignKey) campaignType=\(campaignType) elementId=\(elementId ?? "nil")")

        let event: DigiaExperienceEvent
        switch eventType.lowercased() {
        case "impressed", "viewed": event = .impressed
        case "clicked": event = .clicked(elementID: elementId)
        case "dismissed": event = .dismissed
        default:
            print("[DigiaAnalytics] [DigiaRN] trackEvent: unknown eventType '\(eventType)' — dropping")
            return
        }

        let payload = InAppPayload(
            id: campaignId,
            content: InAppPayloadContent(
                type: campaignType,
                args: [
                    "campaign_id": .string(campaignId),
                    "campaign_key": .string(campaignKey),
                    "campaign_type": .string(campaignType),
                ],
                campaignKey: campaignKey
            )
        )

        Task { @MainActor in
            Digia.captureAnalyticsEvent(event, payload: payload)
        }
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - triggerCampaign

    /// Forwards a campaign payload to the native DigiaCEPDelegate.
    ///
    /// Called by the JS DigiaDelegate.onCampaignTriggered() when a JS CEP
    /// plugin (e.g. DigiaMoEngagePlugin) delivers a campaign. The delegate
    /// routes it into the SwiftUI overlay for rendering.
    @objc
    func triggerCampaign(
        _ id: String,
        content contentMap: NSDictionary,
        cepContext cepContextMap: NSDictionary
    ) {
        let content = buildInAppPayloadContent(from: contentMap)
        let cepContext = (cepContextMap as? [String: String]) ?? [:]
        let payload = InAppPayload(id: id, content: content, cepContext: cepContext)
        print("[DigiaRN] triggerCampaign id=\(id) campaignKey=\(content.campaignKey ?? "nil")")

        Task { @MainActor in
            guard let delegate = self.rnPlugin.delegate else {
                print("[DigiaRN] triggerCampaign: delegate is nil — registerBridge() may not have run yet")
                return
            }
            delegate.onCampaignTriggered(payload)
            Task { @MainActor in
                print("[DigiaRN] triggerCampaign post-call: hasActiveOverlay=\(Digia.hasActiveOverlay)")
            }
        }
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - invalidateCampaign

    @objc
    func invalidateCampaign(_ campaignId: String) {
        Task { @MainActor in
            guard let delegate = self.rnPlugin.delegate else { return }
            delegate.onCampaignInvalidated(campaignId)
        }
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - registerAnchor / unregisterAnchor

    /// Registers a UI element as an anchor point for Guide experiences.
    /// JS sends physical pixels; convert to UIKit points using screen scale.
    @objc
    func registerAnchor(_ key: String, x: Double, y: Double, width: Double, height: Double) {
        let scale = UIScreen.main.scale
        let rect = CGRect(
            x: CGFloat(x) / scale,
            y: CGFloat(y) / scale,
            width: CGFloat(width) / scale,
            height: CGFloat(height) / scale
        )
        Task { @MainActor in
            AnchorRegistry.shared.register(key: key, rect: rect)
        }
    }

    @objc
    func unregisterAnchor(_ key: String) {
        Task { @MainActor in
            AnchorRegistry.shared.unregister(key: key)
        }
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - getRegisteredComponents

    @objc
    func getRegisteredComponents(
        _ resolve: @escaping RCTPromiseResolveBlock,
        reject: @escaping RCTPromiseRejectBlock
    ) {
        resolve([])
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - Internal: mount the SwiftUI overlay host

    /// Mirrors Android's DigiaModule.mountDigiaHost().
    /// Called once after Digia.initialize() succeeds — no need for a manual
    /// <DigiaHostView> anywhere in the JS component tree.
    @MainActor
    private func mountDigiaHost() {
        // Locate the key window's root view controller.
        guard
            let rootVC = UIApplication.shared
                .connectedScenes
                .compactMap({ ($0 as? UIWindowScene)?.keyWindow })
                .first?
                .rootViewController
        else { return }

        // Guard against double-mounting (e.g. fast-refresh).
        if rootVC.view.viewWithTag(Self.overlayMountTag) != nil { return }

        let hc = UIHostingController(rootView: DigiaHostWrapperView())
        let hostView = DigiaRootOverlayView(hostingController: hc)
        hostView.tag = Self.overlayMountTag
        hostView.translatesAutoresizingMaskIntoConstraints = false
        hostView.backgroundColor = .clear

        hc.view.translatesAutoresizingMaskIntoConstraints = false
        hc.view.backgroundColor = .clear

        rootVC.addChild(hc)
        hostView.addSubview(hc.view)
        rootVC.view.addSubview(hostView)
        hc.didMove(toParent: rootVC)

        NSLayoutConstraint.activate([
            hostView.leadingAnchor.constraint(equalTo: rootVC.view.leadingAnchor),
            hostView.trailingAnchor.constraint(equalTo: rootVC.view.trailingAnchor),
            hostView.topAnchor.constraint(equalTo: rootVC.view.topAnchor),
            hostView.bottomAnchor.constraint(equalTo: rootVC.view.bottomAnchor),
            hc.view.leadingAnchor.constraint(equalTo: hostView.leadingAnchor),
            hc.view.trailingAnchor.constraint(equalTo: hostView.trailingAnchor),
            hc.view.topAnchor.constraint(equalTo: hostView.topAnchor),
            hc.view.bottomAnchor.constraint(equalTo: hostView.bottomAnchor),
        ])
    }

    // ────────────────────────────────────────────────────────────────────────
    // MARK: - Private helpers

    private func buildInAppPayloadContent(from map: NSDictionary) -> InAppPayloadContent {
        let pk = map["placementKey"] as? String
        let title = map["title"] as? String
        let text = map["text"] as? String
        let viewId = map["viewId"] as? String
        let command = map["command"] as? String
        let screenId = map["screenId"] as? String
        let campaignKey =
            (map["campaignKey"] as? String) ?? (map["campaign_key"] as? String)
            ?? (map["digia_campaign_key"] as? String)
            ?? (map["digiaKey"] as? String)
        var type = (map["type"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        if type.isEmpty {
            type =
                (pk?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "").isEmpty
                ? "dialog" : "inline"
        }
        let args: [String: JSONValue] = {
            guard let raw = map["args"] as? [String: Any] else { return [:] }
            return raw.compactMapValues { JSONValue(rawValue: $0) }
        }()

        // CEP trigger variables for `{{ }}` interpolation. CleverTap's nudge
        // mapper puts them at `content.variables` (top-level); the command/inline
        // mappers may nest them under `args.variables`. Mirror the JS bridge's
        // `_extractVariables` (content.variables first, then args.variables).
        let variables = Self.variableMap(map["variables"])
            ?? Self.variableMap((map["args"] as? [String: Any])?["variables"])

        return InAppPayloadContent(
            type: type,
            placementKey: pk,
            title: title,
            text: text,
            viewId: viewId,
            command: command,
            args: args,
            screenId: screenId,
            campaignKey: campaignKey,
            variables: variables
        )
    }

    /// Coerces a raw `variables` value into a `[String: String]` map, stringifying
    /// scalar values (string / number / bool) and dropping anything else. Returns
    /// nil for a missing or empty map. Mirrors the JS `parseVariableMap`.
    private static func variableMap(_ raw: Any?) -> [String: String]? {
        guard let dict = raw as? [String: Any] else { return nil }
        var result: [String: String] = [:]
        for (key, value) in dict {
            switch value {
            case let string as String:
                result[key] = string
            case let number as NSNumber:
                // Distinguish a boxed bool from a numeric NSNumber so `true`
                // stays "true" rather than "1".
                if CFGetTypeID(number) == CFBooleanGetTypeID() {
                    result[key] = number.boolValue ? "true" : "false"
                } else {
                    result[key] = number.stringValue
                }
            default:
                continue
            }
        }
        return result.isEmpty ? nil : result
    }
}

// MARK: - DigiaRootOverlayView

/// Full-screen native overlay container for the auto-mounted DigiaHost.
/// It stays touch-transparent when SwiftUI has no active overlay, but lets
/// survey/dialog/sheet content and barriers receive touches when present.
private final class DigiaRootOverlayView: UIView {
    private let hostingController: UIHostingController<DigiaHostWrapperView>

    init(hostingController: UIHostingController<DigiaHostWrapperView>) {
        self.hostingController = hostingController
        super.init(frame: .zero)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
        guard let hcView = hostingController.view else { return nil }

        // When a modal overlay (survey / dialog / bottom sheet / toast / tooltip)
        // is active, the SwiftUI layer must own ALL touches so the dim backdrop
        // both blocks the app underneath and receives barrier-dismiss taps.
        // SwiftUI draws a gesture-only scrim (Color + .onTapGesture) into the
        // hosting view itself with no child UIView, so the `hit === hcView`
        // heuristic below would wrongly treat it as empty and pass the touch
        // through to React Native content.
        let overlayActive = MainActor.assumeIsolated { Digia.hasActiveOverlay }
        if overlayActive {
            return hcView.hitTest(point, with: event) ?? hcView
        }

        let hit = hcView.hitTest(point, with: event)
        if hit == nil || hit === hcView { return nil }
        return hit
    }
}

// MARK: - JSONValue convenience init from Any
extension JSONValue {
    fileprivate init?(rawValue: Any) {
        switch rawValue {
        case let s as String: self = .string(s)
        case let b as Bool: self = .bool(b)
        case let i as Int: self = .int(i)
        case let d as Double: self = .double(d)
        case let arr as [Any]:
            self = .array(arr.compactMap { JSONValue(rawValue: $0) })
        case let dict as [String: Any]:
            let mapped = dict.compactMapValues { JSONValue(rawValue: $0) }
            self = .object(mapped)
        default:
            return nil
        }
    }
}
